src/charm.py

```diff
@@ -260,6 +269,185 @@ def _on_remove(self, event: RemoveEvent):
+            nonce = relation.data[unit].get("nonce")
+            if egress_subnet is None or nonce is None:
+                logger.debug(
+                    "Skipping configuring access for unit %r, egress_subnet or nonce are missing",
+                    unit.name,
+                )
+                continue
+            nonces.append(nonce)
+            self._ensure_unit_credentials(vault, relation, unit.name, mount, nonce, egress_subnet)
+
+        # Remove any stale nonce
+        credential_nonces = self.vault_kv.get_credentials(relation).keys()
+        stale_nonces = list(set(credential_nonces) - set(nonces))
+        self.vault_kv.remove_unit_credentials(relation, stale_nonces)
+
+    def _ensure_unit_credentials(
+        self,
+        vault: Vault,
+        relation: Relation,
+        unit_name: str,
+        mount: str,
+        nonce: str,
+        egress_subnet: str,
+    ):
+        """Ensures a unit has credentials to access the vault-kv mount."""
+        policy_name = role_name = mount + "-" + unit_name.replace("/", "-")
+        vault.configure_kv_policy(policy_name, mount)
+        role_id = vault.configure_approle(role_name, [egress_subnet], [policy_name])
+        secret = self._create_or_update_kv_secret(
+            vault,
+            relation,
+            role_id,
+            role_name,
+            egress_subnet,
+        )
+        self.vault_kv.set_unit_credentials(relation, nonce, secret)
+
+    def _create_or_update_kv_secret(
+        self,
+        vault: Vault,
+        relation: Relation,
+        role_id: str,
+        role_name: str,
+        egress_subnet: str,
+    ) -> Secret:
+        """Create or update a KV secret for a unit.
+
+        Fetch secret id from peer relation, if it exists, update the secret,
+        otherwise create it.
+        """
+        label = KV_SECRET_PREFIX + role_name
+        secret_id = self._get_vault_kv_secret_in_peer_relation(label)
+        if secret_id is None:
+            return self._create_kv_secret(
+                vault, relation, role_id, role_name, egress_subnet, label
+            )
+        else:
+            return self._update_kv_secret(
+                vault, relation, role_name, egress_subnet, label, secret_id
+            )
+
+    def _create_kv_secret(
+        self,
+        vault: Vault,
+        relation: Relation,
+        role_id: str,
+        role_name: str,
+        egress_subnet: str,
+        label: str,
+    ) -> Secret:
+        """Create a vault kv secret, store its id in the peer relation and return it."""
+        role_secret_id = vault.generate_role_secret_id(role_name, [egress_subnet])
+        secret = self.app.add_secret(
+            {"role-id": role_id, "role-secret-id": role_secret_id},
+            label=label,
+        )
+        if secret.id is None:
+            raise RuntimeError(f"Unexpected error, just created secret {label!r} has no id")
+        self._set_vault_kv_secret_in_peer_relation(label, secret.id)
```

Comment:
Why do we store it in the peer relation data? Do other peers need access to it?

Reply:
Since credentials are stored per nonce, and this can change on a pod reschedule, we need somewhere to store the reference between label and ID.

(normally, we should be able to get the secret from the label, and fill in the id, but this fills the short form of the secret, which is not compatible with CMRs)

Reply:
By fill in the ID, I'm talking about the `secret-info-get` hook. But this hook does not return an ID that can be used in CMRs.